### PR TITLE
[FLINK-16669][python][table] Support Python UDF in SQL function DDL.

### PR DESCRIPTION
**Pull request by WeiZhong94**
## What is the purpose of the change

*This pull request supports Python UDF in SQL function DDL.*

## Brief change log

  - *Refactor `FunctionCatalog` and `TableEnvironment` to instantiate function as late as possible.*
  - *Add 'PYTHON' keyword to sql parser to support Python UDF SQL DDL.*
  - *Add `PythonFunctionFactory` to request the Python UDF metadata from python process.*

## Verifying this change
This change is already covered by existing tests, such as *FunctionCatalogTest*, *PythonFunctionSqlTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

### CODE REVIEW

File: flink-python/src/test/java/org/apache/flink/client/python/PythonFunctionSqlTest.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.UUID;
+
+import static org.apache.flink.python.PythonOptions.PYTHON_FILES;
+
+/**
+ * Tests for PythonFunctionSqlTest.
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/client/python/PythonFunctionSqlTest.java:42]: correct the comments.

File: flink-python/src/test/java/org/apache/flink/client/python/PythonFunctionSqlTest.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ */
+public class PythonFunctionSqlTest {
+
+	private String tmpdir = "";
+	private BatchTableEnvironment flinkTableEnv;
+	private StreamTableEnvironment blinkTableEnv;
+	private Table flinkSourceTable;
+	private Table blinkSourceTable;
+
+	private void closeStartedPythonProcess() {
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/client/python/PythonFunctionSqlTest.java:52]: Move the private method to the bottom of the class

File: flink-python/src/test/java/org/apache/flink/client/python/PythonFunctionSqlTest.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+				PythonFunctionFactoryUtil.pythonProcessShutdownHook.run();
+				Runtime.getRuntime().removeShutdownHook(PythonFunctionFactoryUtil.pythonProcessShutdownHook);
+				PythonFunctionFactoryUtil.pythonProcessShutdownHook = null;
+			}
+		}
+	}
+
+	@Before
+	public void prepareEnvironment() throws IOException {
+		closeStartedPythonProcess();
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/client/python/PythonFunctionSqlTest.java:64]: Could we move closeStartedPythonProcess to a method marked as `@AfterClass`?

File: flink-python/src/test/java/org/apache/flink/client/python/PythonFunctionSqlTest.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.UUID;
+
+import static org.apache.flink.python.PythonOptions.PYTHON_FILES;
+
+/**
+ * Tests for PythonFunctionSqlTest.
+ */
+public class PythonFunctionSqlTest {
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/client/python/PythonFunctionSqlTest.java:44]: Rename it to PythonFunctionFactoryTest

File: flink-python/src/test/java/org/apache/flink/client/python/PythonFunctionSqlTest.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+		FileUtils.deleteDirectory(new File(tmpdir));
+	}
+
+	@Test
+	public void testPythonFunctionFactory() {
+		// flink temporary catalog
+		flinkTableEnv.sqlUpdate("create temporary function func1 as 'test1.func1' language python");
+		Table flinkTable = flinkSourceTable.select("func1(str)");
+		String flinkPlan = flinkTableEnv.explain(flinkTable);
+		Assert.assertTrue(flinkPlan.contains("DataSetPythonCalc(select=[func1(f0) AS _c0])"));
```
Comment (dianfu) [inline flink-python/src/test/java/org/apache/flink/client/python/PythonFunctionSqlTest.java:100]: Change `DataSetPythonCalc` to `PythonCalc` and refactor the test case a bit to remove the code duplication?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 6]
```diff
@@ -945,73 +941,6 @@ private void dropSystemFunction(DropTempSystemFunctionOperation operation) {
 		}
 	}
 
-	private <T, ACC> void  registerCatalogFunctionInFunctionCatalog(
```
Comment (dianfu) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:948]: Could you split this into a separate commit?

File: flink-python/src/main/java/org/apache/flink/client/python/PythonDriver.java [hunk 7]
```diff
@@ -67,20 +69,22 @@ public static void main(String[] args) {
...
-			pythonEnv.systemEnv.put("PYFLINK_GATEWAY_PORT", String.valueOf(gatewayServer.getListeningPort()));
 			// start the python process.
-			Process pythonProcess = PythonDriverEnvUtils.startPythonProcess(pythonEnv, commands);
+			Process pythonProcess = PythonEnvUtils.launchPy4jPythonClient(
+				gatewayServer,
+				config,
+				commands,
+				pythonDriverOptions.getEntryPointScript().orElse(null),
+				tmpDir);
+			setGatewayServer(gatewayServer);
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-python/src/main/java/org/apache/flink/client/python/PythonDriver.java]: As the `PythonDriver`, `PythonGatewayServer`, `PythonFunctionFactory` will share the same gatewayServer object, it is necessary to set the gatewayServer to a static variable here.

File: flink-python/src/main/java/org/apache/flink/client/python/PythonEnvUtils.java [hunk 8]
```diff
@@ -247,4 +257,90 @@ static Process startPythonProcess(PythonEnvironment pythonEnv, List<String> comm
...
+		Thread thread = new Thread(() -> {
+			int freePort;
+			try {
+				freePort = findFreePort();
+			} catch (IOException e) {
+				throw new RuntimeException("Could not find a free port for Py4jCallbackClient.");
+			}
+			GatewayServer server = new GatewayServer.GatewayServerBuilder()
+				.gateway(new Gateway(new ConcurrentHashMap<String, Object>(), new CallbackClient(freePort)))
+				.javaPort(0)
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-python/src/main/java/org/apache/flink/client/python/PythonEnvUtils.java:281]: Because the loopback address is the default value of the `GatewayServer`.

File: flink-python/src/main/java/org/apache/flink/client/python/PythonFunctionFactoryUtil.java [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+	@VisibleForTesting
+	static Thread pythonProcessShutdownHook = null;
+
+	public static synchronized PythonFunctionFactory getPythonFunctionFactory(ReadableConfig config)
+			throws IOException, ExecutionException, InterruptedException {
+		if (pythonFunctionFactory != null) {
+			return pythonFunctionFactory;
+		} else {
+			if (getGatewayServer() == null) {
```
[the remark this answers is not in the document]
Comment (WeiZhong94) [reply flink-python/src/main/java/org/apache/flink/client/python/PythonFunctionFactoryUtil.java]: If the entry point of the Java process is `PythonDriver` or `PythonGatewayServer`, it is possible to hit this situation.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have addressed your comments in the latest commits.

**Comment by WeiZhong94** [comment]
@dianfu Thanks for your review! I have addressed your comments in the latest commit.

**Comment by dianfu** [comment]
@WeiZhong94 Thanks a lot for the update. Will merge after the travis/azure passed.
